src/librustc/hir/mod.rs

```diff
@@ -2489,6 +2489,29 @@ bitflags! {
         /// #[used], indicates that LLVM can't eliminate this function (but the
         /// linker can!)
         const USED                      = 1 << 9;
+        /// #[ffi_pure], indicates that a function has no effects except for
```

Comment:
Is non-termination an effect? If so, is non-termination deterministic?

Reply:
> Is non-termination an effect?

In C and C++, I think so, calling an `#[ffi_pure]` function that returns `-> !` is UB.

Reply:
I didn't expect that so clarifying the comment to that end would be helpful.

Reply:
> I didn't expect that so clarifying the comment to that end would be helpful.

I think this becomes clearer when one states what a `#[ffi_pure] -> !` function "could" do: 

* no-termination: loop forever, ...
* terminate: panic / abort / throw / ...

Since `#[ffi_pure]` prevents side-effects, and:

*  infinite loops without side-effects are undefined behavior in C, so if this function never terminates, calling it is UB anyways.

* termination of any form (`panic`, `abort`, `throw`, etc.) is definitely a side-effect in C (and also in Rust), if this function terminates, calling it is also UB. 

That is, independently of what the programmer intended such a function to do (terminate or not terminate), calling it is always UB.

Reply:
> * terminate: panic / abort / throw / ...

No function of type `-> !` could possibly terminate; they will always diverge. Panicing, aborting, and throwing are all forms of divergence.

> Since `#[ffi_pure]` prevents side-effects, and:
> 
>     * infinite loops without side-effects are undefined behavior in C, so if this function never terminates, calling it is UB anyways.
> 
>     * termination of any form (`panic`, `abort`, `throw`, etc.) is definitely a side-effect in C (and also in Rust), so if this function terminates, calling it is also UB.
> 
> 
> That is, independently of what the programmer intended such a C function to do (terminate or not terminate), calling it is always UB.
> 
> EDIT: In Rust, an infinite loop without side-effects is not UB, so for FFI with Rust code this attribute should probably do something else.

Such a discussion would be useful in the doc comments.

Reply:
> > Is non-termination an effect?
>
> In C and C++, I think so, calling an #[ffi_pure] function that returns -> ! is UB.

It is in C, but not C++. This is one of the reasons why we have issues with `loop{}` being a soundness hole currently.

Reply:
I think this is also the case in C++. [cppreference: C++ memory model - Progress guarantee](https://en.cppreference.com/w/cpp/language/memory_model#Progress_guarantee
) and the latest draft of the C++ standard [intro.progress.1](http://eel.is/c++draft/intro.progress#1) call this out:

> The implementation may assume that any thread will eventually do one of the following:
> - (1.1) terminate,
> - (1.2) make a call to a library I/O function,
> - (1.3) perform an access through a volatile glvalue, or
> - (1.4) perform a synchronization operation or an atomic operation.
>
> [ Note: This is intended to allow compiler transformations such as removal of empty loops, even when termination cannot be proven. — end note ]

Reply:
“It is [an effect] in C, but not in C++” is what I meant, rather than “it is [UB]...”.

Reply:
We discussed this on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/termination.20c.20vs.20c.2B.2B). We aren't 100% sure, but we believe that in both C and C++ calling a `noreturn,pure` function is undefined behavior. 

In both C and C++, a `noreturn` function must either diverge, or have an infinite loop with side-effects, but side-effects are not allowed by the `pure` attribute. 

Implementing such a function using an infinite loop without side-effects would lead to UB in both C and C++. In C++, infinite loops without side-effects are UB. They are not UB in C, but in C they can be assumed to terminate, which would result in the `noreturn` function returning, which would be UB.